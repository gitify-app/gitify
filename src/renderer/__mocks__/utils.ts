import * as helpers from '../utils/helpers';

export function mockDirectoryPath() {
  jest.spyOn(helpers, 'getDirectoryPath').mockReturnValue('/mocked/dir/name');
}

/**
 * Ensure stable snapshots for our randomized emoji use-cases
 */
export function ensureStableEmojis() {
  global.Math.random = jest.fn(() => 0.1);
}
