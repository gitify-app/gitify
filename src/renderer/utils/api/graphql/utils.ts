export function formatAsGitHubSearchSyntax(
  repo: string,
  title: string,
): string {
  return `${title} in:title repo:${repo}`;
}
