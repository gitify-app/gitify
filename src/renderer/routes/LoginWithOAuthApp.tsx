import { type FC, useCallback, useContext } from 'react';
import { Form, type FormRenderProps } from 'react-final-form';
import { useNavigate } from 'react-router-dom';

import { BookIcon, PersonIcon, SignInIcon } from '@primer/octicons-react';
import { Button, Stack, Text, Tooltip } from '@primer/react';

import { logError } from '../../shared/logger';
import { FieldInput } from '../components/fields/FieldInput';
import { Header } from '../components/primitives/Header';
import { AppContext } from '../context/App';
import type { ClientID, ClientSecret, Hostname, Token } from '../types';
import type { LoginOAuthAppOptions } from '../utils/auth/types';
import {
  getNewOAuthAppURL,
  isValidClientId,
  isValidHostname,
  isValidToken,
} from '../utils/auth/utils';
import { openExternalLink } from '../utils/comms';
import { Constants } from '../utils/constants';

interface IValues {
  hostname?: Hostname;
  clientId?: ClientID;
  clientSecret?: ClientSecret;
}

interface IFormErrors {
  hostname?: string;
  clientId?: string;
  clientSecret?: string;
}

export const validate = (values: IValues): IFormErrors => {
  const errors: IFormErrors = {};

  if (!values.hostname) {
    errors.hostname = 'Required';
  } else if (!isValidHostname(values.hostname)) {
    errors.hostname = 'Invalid hostname.';
  }

  if (!values.clientId) {
    errors.clientId = 'Required';
  } else if (!isValidClientId(values.clientId)) {
    errors.clientId = 'Invalid client id.';
  }

  if (!values.clientSecret) {
    errors.clientSecret = 'Required';
  } else if (!isValidToken(values.clientSecret as unknown as Token)) {
    errors.clientSecret = 'Invalid client secret.';
  }

  return errors;
};

export const LoginWithOAuthAppRoute: FC = () => {
  const navigate = useNavigate();

  const { loginWithOAuthApp } = useContext(AppContext);

  const renderForm = (formProps: FormRenderProps) => {
    const { handleSubmit, submitting, pristine, values } = formProps;

    return (
      <form onSubmit={handleSubmit}>
        <FieldInput
          name="hostname"
          label="Hostname"
          placeholder="github.company.com"
          helpText={
            <Stack direction="vertical" gap="condensed">
              <Stack direction="horizontal" align="center" gap="condensed">
                <Button
                  size="small"
                  leadingVisual={PersonIcon}
                  disabled={!values.hostname}
                  onClick={() =>
                    openExternalLink(getNewOAuthAppURL(values.hostname))
                  }
                  data-testid="login-create-oauth-app"
                >
                  Create new OAuth App
                </Button>
                <Text>on GitHub then paste your</Text>
              </Stack>
              <Text>
                <Text as="i">client id and client secret</Text> below.
              </Text>
            </Stack>
          }
        />

        <FieldInput name="clientId" label="Client ID" placeholder="123456789" />

        <FieldInput
          name="clientSecret"
          label="Client Secret"
          placeholder="ABC123DEF456"
        />

        <Stack direction="horizontal" justify="space-between" align="center">
          <Tooltip text="GitHub documentation">
            <Button
              size="small"
              leadingVisual={BookIcon}
              onClick={() => openExternalLink(Constants.GITHUB_DOCS.OAUTH_URL)}
              data-testid="login-docs"
            >
              Docs
            </Button>
          </Tooltip>

          <Tooltip text="Login">
            <Button
              variant="primary"
              leadingVisual={SignInIcon}
              disabled={submitting || pristine}
              type="submit"
              data-testid="login-submit"
            >
              Login
            </Button>
          </Tooltip>
        </Stack>
      </form>
    );
  };

  const login = useCallback(
    async (data: IValues) => {
      try {
        await loginWithOAuthApp(data as LoginOAuthAppOptions);
        navigate(-1);
      } catch (err) {
        logError('loginWithOAuthApp', 'Failed to login with OAuth App', err);
      }
    },
    [loginWithOAuthApp],
  );

  return (
    <div>
      <Header icon={PersonIcon}>Login with OAuth App</Header>
      <div className="px-8">
        <Form
          initialValues={{
            hostname: '',
            clientId: '',
            clientSecret: '',
          }}
          onSubmit={login}
          validate={validate}
        >
          {renderForm}
        </Form>
      </div>
    </div>
  );
};
