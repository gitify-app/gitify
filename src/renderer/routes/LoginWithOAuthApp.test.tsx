import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { AppContext } from '../context/App';
import type { AuthState, ClientID, ClientSecret, Hostname } from '../types';
import * as comms from '../utils/comms';
import { LoginWithOAuthAppRoute, validateForm } from './LoginWithOAuthApp';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('renderer/routes/LoginWithOAuthApp.tsx', () => {
  const mockLoginWithOAuthApp = jest.fn();

  const openExternalLinkMock = jest
    .spyOn(comms, 'openExternalLink')
    .mockImplementation();

  const mockAuth: AuthState = {
    accounts: [],
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const tree = render(
      <AppContext.Provider value={{ auth: mockAuth }}>
        <MemoryRouter>
          <LoginWithOAuthAppRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('let us go back', () => {
    render(
      <AppContext.Provider value={{ auth: mockAuth }}>
        <MemoryRouter>
          <LoginWithOAuthAppRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.click(screen.getByTestId('header-nav-back'));

    expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
  });

  it('should validate the form values', () => {
    const emptyValues = {
      hostname: null,
      clientId: null,
      clientSecret: null,
    };

    let values = {
      ...emptyValues,
    };
    expect(validateForm(values).hostname).toBe('Hostname is required');
    expect(validateForm(values).clientId).toBe('Client ID is required');
    expect(validateForm(values).clientSecret).toBe('Client Secret is required');

    values = {
      ...emptyValues,
      hostname: 'hello' as Hostname,
      clientId: '!@£INVALID-.1' as ClientID,
      clientSecret: '!@£INVALID-.1' as ClientSecret,
    };
    expect(validateForm(values).hostname).toBe('Hostname format is invalid');
    expect(validateForm(values).clientId).toBe('Client ID format is invalid');
    expect(validateForm(values).clientSecret).toBe(
      'Client Secret format is invalid',
    );
  });

  describe("'Create new OAuth App' button", () => {
    it('should be disabled if no hostname configured', async () => {
      render(
        <AppContext.Provider value={{ auth: mockAuth }}>
          <MemoryRouter>
            <LoginWithOAuthAppRoute />
          </MemoryRouter>
        </AppContext.Provider>,
      );

      fireEvent.change(screen.getByTestId('login-hostname'), {
        target: { value: '' },
      });

      fireEvent.click(screen.getByTestId('login-create-oauth-app'));

      expect(openExternalLinkMock).toHaveBeenCalledTimes(0);
    });

    it('should open in browser if hostname configured', async () => {
      render(
        <AppContext.Provider value={{ auth: mockAuth }}>
          <MemoryRouter>
            <LoginWithOAuthAppRoute />
          </MemoryRouter>
        </AppContext.Provider>,
      );

      fireEvent.change(screen.getByTestId('login-hostname'), {
        target: { value: 'company.github.com' },
      });

      fireEvent.click(screen.getByTestId('login-create-oauth-app'));

      expect(openExternalLinkMock).toHaveBeenCalledTimes(1);
    });
  });

  it('should login using a token - success', async () => {
    mockLoginWithOAuthApp.mockResolvedValueOnce(null);

    render(
      <AppContext.Provider
        value={{
          loginWithOAuthApp: mockLoginWithOAuthApp,
        }}
      >
        <MemoryRouter>
          <LoginWithOAuthAppRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.change(screen.getByTestId('login-hostname'), {
      target: { value: 'github.com' },
    });
    fireEvent.change(screen.getByTestId('login-clientId'), {
      target: { value: '1234567890_ASDFGHJKL' },
    });
    fireEvent.change(screen.getByTestId('login-clientSecret'), {
      target: { value: '1234567890_asdfghjklPOIUYTREWQ0987654321' },
    });

    fireEvent.click(screen.getByTestId('login-submit'));

    await waitFor(() => expect(mockLoginWithOAuthApp).toHaveBeenCalledTimes(1));

    expect(mockLoginWithOAuthApp).toHaveBeenCalledTimes(1);
    expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
  });

  it('should render the form with errors', () => {
    render(
      <AppContext.Provider value={{ auth: mockAuth }}>
        <MemoryRouter>
          <LoginWithOAuthAppRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.change(screen.getByTestId('login-hostname'), {
      target: { value: 'test' },
    });
    fireEvent.change(screen.getByTestId('login-clientId'), {
      target: { value: '123' },
    });
    fireEvent.change(screen.getByTestId('login-clientSecret'), {
      target: { value: 'abc' },
    });

    fireEvent.click(screen.getByTestId('login-submit'));

    expect(screen.getByTestId('login-errors')).toBeTruthy();
    expect(screen.getByText('Hostname format is invalid')).toBeTruthy();
    expect(screen.getByText('Client ID format is invalid')).toBeTruthy();
    expect(screen.getByText('Client Secret format is invalid')).toBeTruthy();
  });

  it('should open help docs in the browser', async () => {
    render(
      <AppContext.Provider value={{ auth: mockAuth }}>
        <MemoryRouter>
          <LoginWithOAuthAppRoute />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.click(screen.getByTestId('login-docs'));

    expect(openExternalLinkMock).toHaveBeenCalledTimes(1);
  });
});
