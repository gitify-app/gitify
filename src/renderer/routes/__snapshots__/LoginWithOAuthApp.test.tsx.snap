// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer/routes/LoginWithOAuthApp.tsx renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
          data-align="stretch"
          data-direction="horizontal"
          data-justify="space-between"
          data-padding="spacious"
          data-wrap="nowrap"
        >
          <button
            aria-describedby=":r0:-loading-announcement"
            aria-labelledby="Go Back"
            class="types__StyledButton-sc-ws60qy-0 kQRupk"
            data-component="IconButton"
            data-loading="false"
            data-no-visuals="true"
            data-testid="header-nav-back"
            sx="[object Object]"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="octicon octicon-arrow-left"
              display="inline-block"
              fill="currentColor"
              focusable="false"
              height="16"
              overflow="visible"
              style="vertical-align: text-bottom;"
              viewBox="0 0 16 16"
              width="16"
            >
              <path
                d="M7.78 12.53a.75.75 0 0 1-1.06 0L2.47 8.28a.75.75 0 0 1 0-1.06l4.25-4.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042L4.81 7h7.44a.75.75 0 0 1 0 1.5H4.81l2.97 2.97a.75.75 0 0 1 0 1.06Z"
              />
            </svg>
          </button>
          <legend>
            <div
              class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
              data-align="center"
              data-direction="horizontal"
              data-gap="condensed"
              data-justify="start"
              data-padding="none"
              data-wrap="nowrap"
              id="login-with oauth app"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-person"
                display="inline-block"
                fill="currentColor"
                focusable="false"
                height="16"
                overflow="visible"
                style="vertical-align: text-bottom;"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
                />
              </svg>
              <h2
                class="Heading__StyledHeading-sc-1c1dgg0-0 fuhZuh"
                sx="[object Object]"
              >
                Login with OAuth App
              </h2>
            </div>
          </legend>
        </div>
        <div
          class="px-8"
        >
          <div
            class="flex flex-col h-screen"
            data-testid="Login With OAuth App"
          >
            <form>
              <div
                class="mb-4"
              >
                <label
                  class="mb-2 block text-sm font-semibold tracking-wide"
                  for="hostname"
                >
                  Hostname
                </label>
                <input
                  class="text-sm mb-2 px-4 py-1.5 block w-full appearance-none rounded border bg-gitify-input-rest focus:bg-gitify-input-focus focus:outline-none border-red-500"
                  id="hostname"
                  name="hostname"
                  placeholder="github.company.com"
                  type="text"
                  value=""
                />
                <div
                  class="mt-3 text-xs text-gitify-font"
                >
                  <div
                    class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
                    data-align="stretch"
                    data-direction="vertical"
                    data-gap="condensed"
                    data-justify="start"
                    data-padding="none"
                    data-wrap="nowrap"
                  >
                    <div
                      class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
                      data-align="center"
                      data-direction="horizontal"
                      data-gap="condensed"
                      data-justify="start"
                      data-padding="none"
                      data-wrap="nowrap"
                    >
                      <button
                        aria-describedby=":r1:-loading-announcement"
                        class="types__StyledButton-sc-ws60qy-0 jTCgqj"
                        data-loading="false"
                        data-size="small"
                        data-testid="login-create-oauth-app"
                        disabled=""
                        sx="[object Object]"
                        type="button"
                      >
                        <span
                          class="Box-sc-g0xbh4-0 srpGE"
                          data-component="buttonContent"
                          sx="[object Object]"
                        >
                          <span
                            class="Box-sc-g0xbh4-0 mhHOW"
                            data-component="leadingVisual"
                            sx="[object Object]"
                          >
                            <svg
                              aria-hidden="true"
                              class="octicon octicon-person"
                              display="inline-block"
                              fill="currentColor"
                              focusable="false"
                              height="16"
                              overflow="visible"
                              style="vertical-align: text-bottom;"
                              viewBox="0 0 16 16"
                              width="16"
                            >
                              <path
                                d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
                              />
                            </svg>
                          </span>
                          <span
                            data-component="text"
                          >
                            Create new OAuth App
                          </span>
                        </span>
                      </button>
                      <span
                        class="Text-sc-17v1xeu-0 hmAFGJ"
                      >
                        on GitHub then paste your
                      </span>
                    </div>
                    <span
                      class="Text-sc-17v1xeu-0 hmAFGJ"
                    >
                      <i
                        class="Text-sc-17v1xeu-0 hmAFGJ"
                      >
                        client id and client secret
                      </i>
                       below.
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="mb-4"
              >
                <label
                  class="mb-2 block text-sm font-semibold tracking-wide"
                  for="clientId"
                >
                  Client ID
                </label>
                <input
                  class="text-sm mb-2 px-4 py-1.5 block w-full appearance-none rounded border bg-gitify-input-rest focus:bg-gitify-input-focus focus:outline-none border-red-500"
                  id="clientId"
                  name="clientId"
                  placeholder="123456789"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="mb-4"
              >
                <label
                  class="mb-2 block text-sm font-semibold tracking-wide"
                  for="clientSecret"
                >
                  Client Secret
                </label>
                <input
                  class="text-sm mb-2 px-4 py-1.5 block w-full appearance-none rounded border bg-gitify-input-rest focus:bg-gitify-input-focus focus:outline-none border-red-500"
                  id="clientSecret"
                  name="clientSecret"
                  placeholder="ABC123DEF456"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
                data-align="center"
                data-direction="horizontal"
                data-justify="space-between"
                data-padding="none"
                data-wrap="nowrap"
              >
                <span
                  aria-label="GitHub documentation"
                  class="Tooltip__TooltipBase-sc-17tf59c-0 fcVRmg tooltipped-n"
                  id=":r2:"
                  role="tooltip"
                >
                  <button
                    aria-describedby=":r3:-loading-announcement"
                    class="types__StyledButton-sc-ws60qy-0 jTCgqj"
                    data-loading="false"
                    data-size="small"
                    data-testid="login-docs"
                    sx="[object Object]"
                    type="button"
                  >
                    <span
                      class="Box-sc-g0xbh4-0 srpGE"
                      data-component="buttonContent"
                      sx="[object Object]"
                    >
                      <span
                        class="Box-sc-g0xbh4-0 mhHOW"
                        data-component="leadingVisual"
                        sx="[object Object]"
                      >
                        <svg
                          aria-hidden="true"
                          class="octicon octicon-book"
                          display="inline-block"
                          fill="currentColor"
                          focusable="false"
                          height="16"
                          overflow="visible"
                          style="vertical-align: text-bottom;"
                          viewBox="0 0 16 16"
                          width="16"
                        >
                          <path
                            d="M0 1.75A.75.75 0 0 1 .75 1h4.253c1.227 0 2.317.59 3 1.501A3.743 3.743 0 0 1 11.006 1h4.245a.75.75 0 0 1 .75.75v10.5a.75.75 0 0 1-.75.75h-4.507a2.25 2.25 0 0 0-1.591.659l-.622.621a.75.75 0 0 1-1.06 0l-.622-.621A2.25 2.25 0 0 0 5.258 13H.75a.75.75 0 0 1-.75-.75Zm7.251 10.324.004-5.073-.002-2.253A2.25 2.25 0 0 0 5.003 2.5H1.5v9h3.757a3.75 3.75 0 0 1 1.994.574ZM8.755 4.75l-.004 7.322a3.752 3.752 0 0 1 1.992-.572H14.5v-9h-3.495a2.25 2.25 0 0 0-2.25 2.25Z"
                          />
                        </svg>
                      </span>
                      <span
                        data-component="text"
                      >
                        Docs
                      </span>
                    </span>
                  </button>
                </span>
                <span
                  aria-label="Login"
                  class="Tooltip__TooltipBase-sc-17tf59c-0 fcVRmg tooltipped-n"
                  id=":r4:"
                  role="tooltip"
                >
                  <button
                    aria-describedby=":r5:-loading-announcement"
                    class="types__StyledButton-sc-ws60qy-0 gEWhIh"
                    data-loading="false"
                    data-testid="login-submit"
                    disabled=""
                    sx="[object Object]"
                    type="submit"
                  >
                    <span
                      class="Box-sc-g0xbh4-0 srpGE"
                      data-component="buttonContent"
                      sx="[object Object]"
                    >
                      <span
                        class="Box-sc-g0xbh4-0 mhHOW"
                        data-component="leadingVisual"
                        sx="[object Object]"
                      >
                        <svg
                          aria-hidden="true"
                          class="octicon octicon-sign-in"
                          display="inline-block"
                          fill="currentColor"
                          focusable="false"
                          height="16"
                          overflow="visible"
                          style="vertical-align: text-bottom;"
                          viewBox="0 0 16 16"
                          width="16"
                        >
                          <path
                            d="M2 2.75C2 1.784 2.784 1 3.75 1h2.5a.75.75 0 0 1 0 1.5h-2.5a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h2.5a.75.75 0 0 1 0 1.5h-2.5A1.75 1.75 0 0 1 2 13.25Zm6.56 4.5h5.69a.75.75 0 0 1 0 1.5H8.56l1.97 1.97a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L6.22 8.53a.75.75 0 0 1 0-1.06l3.25-3.25a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734Z"
                          />
                        </svg>
                      </span>
                      <span
                        data-component="text"
                      >
                        Login
                      </span>
                    </span>
                  </button>
                </span>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
        data-align="stretch"
        data-direction="horizontal"
        data-justify="space-between"
        data-padding="spacious"
        data-wrap="nowrap"
      >
        <button
          aria-describedby=":r0:-loading-announcement"
          aria-labelledby="Go Back"
          class="types__StyledButton-sc-ws60qy-0 kQRupk"
          data-component="IconButton"
          data-loading="false"
          data-no-visuals="true"
          data-testid="header-nav-back"
          sx="[object Object]"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-arrow-left"
            display="inline-block"
            fill="currentColor"
            focusable="false"
            height="16"
            overflow="visible"
            style="vertical-align: text-bottom;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M7.78 12.53a.75.75 0 0 1-1.06 0L2.47 8.28a.75.75 0 0 1 0-1.06l4.25-4.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042L4.81 7h7.44a.75.75 0 0 1 0 1.5H4.81l2.97 2.97a.75.75 0 0 1 0 1.06Z"
            />
          </svg>
        </button>
        <legend>
          <div
            class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
            data-align="center"
            data-direction="horizontal"
            data-gap="condensed"
            data-justify="start"
            data-padding="none"
            data-wrap="nowrap"
            id="login-with oauth app"
          >
            <svg
              aria-hidden="true"
              class="octicon octicon-person"
              display="inline-block"
              fill="currentColor"
              focusable="false"
              height="16"
              overflow="visible"
              style="vertical-align: text-bottom;"
              viewBox="0 0 16 16"
              width="16"
            >
              <path
                d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
              />
            </svg>
            <h2
              class="Heading__StyledHeading-sc-1c1dgg0-0 fuhZuh"
              sx="[object Object]"
            >
              Login with OAuth App
            </h2>
          </div>
        </legend>
      </div>
      <div
        class="px-8"
      >
        <div
          class="flex flex-col h-screen"
          data-testid="Login With OAuth App"
        >
          <form>
            <div
              class="mb-4"
            >
              <label
                class="mb-2 block text-sm font-semibold tracking-wide"
                for="hostname"
              >
                Hostname
              </label>
              <input
                class="text-sm mb-2 px-4 py-1.5 block w-full appearance-none rounded border bg-gitify-input-rest focus:bg-gitify-input-focus focus:outline-none border-red-500"
                id="hostname"
                name="hostname"
                placeholder="github.company.com"
                type="text"
                value=""
              />
              <div
                class="mt-3 text-xs text-gitify-font"
              >
                <div
                  class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
                  data-align="stretch"
                  data-direction="vertical"
                  data-gap="condensed"
                  data-justify="start"
                  data-padding="none"
                  data-wrap="nowrap"
                >
                  <div
                    class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
                    data-align="center"
                    data-direction="horizontal"
                    data-gap="condensed"
                    data-justify="start"
                    data-padding="none"
                    data-wrap="nowrap"
                  >
                    <button
                      aria-describedby=":r1:-loading-announcement"
                      class="types__StyledButton-sc-ws60qy-0 jTCgqj"
                      data-loading="false"
                      data-size="small"
                      data-testid="login-create-oauth-app"
                      disabled=""
                      sx="[object Object]"
                      type="button"
                    >
                      <span
                        class="Box-sc-g0xbh4-0 srpGE"
                        data-component="buttonContent"
                        sx="[object Object]"
                      >
                        <span
                          class="Box-sc-g0xbh4-0 mhHOW"
                          data-component="leadingVisual"
                          sx="[object Object]"
                        >
                          <svg
                            aria-hidden="true"
                            class="octicon octicon-person"
                            display="inline-block"
                            fill="currentColor"
                            focusable="false"
                            height="16"
                            overflow="visible"
                            style="vertical-align: text-bottom;"
                            viewBox="0 0 16 16"
                            width="16"
                          >
                            <path
                              d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
                            />
                          </svg>
                        </span>
                        <span
                          data-component="text"
                        >
                          Create new OAuth App
                        </span>
                      </span>
                    </button>
                    <span
                      class="Text-sc-17v1xeu-0 hmAFGJ"
                    >
                      on GitHub then paste your
                    </span>
                  </div>
                  <span
                    class="Text-sc-17v1xeu-0 hmAFGJ"
                  >
                    <i
                      class="Text-sc-17v1xeu-0 hmAFGJ"
                    >
                      client id and client secret
                    </i>
                     below.
                  </span>
                </div>
              </div>
            </div>
            <div
              class="mb-4"
            >
              <label
                class="mb-2 block text-sm font-semibold tracking-wide"
                for="clientId"
              >
                Client ID
              </label>
              <input
                class="text-sm mb-2 px-4 py-1.5 block w-full appearance-none rounded border bg-gitify-input-rest focus:bg-gitify-input-focus focus:outline-none border-red-500"
                id="clientId"
                name="clientId"
                placeholder="123456789"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-4"
            >
              <label
                class="mb-2 block text-sm font-semibold tracking-wide"
                for="clientSecret"
              >
                Client Secret
              </label>
              <input
                class="text-sm mb-2 px-4 py-1.5 block w-full appearance-none rounded border bg-gitify-input-rest focus:bg-gitify-input-focus focus:outline-none border-red-500"
                id="clientSecret"
                name="clientSecret"
                placeholder="ABC123DEF456"
                type="text"
                value=""
              />
            </div>
            <div
              class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
              data-align="center"
              data-direction="horizontal"
              data-justify="space-between"
              data-padding="none"
              data-wrap="nowrap"
            >
              <span
                aria-label="GitHub documentation"
                class="Tooltip__TooltipBase-sc-17tf59c-0 fcVRmg tooltipped-n"
                id=":r2:"
                role="tooltip"
              >
                <button
                  aria-describedby=":r3:-loading-announcement"
                  class="types__StyledButton-sc-ws60qy-0 jTCgqj"
                  data-loading="false"
                  data-size="small"
                  data-testid="login-docs"
                  sx="[object Object]"
                  type="button"
                >
                  <span
                    class="Box-sc-g0xbh4-0 srpGE"
                    data-component="buttonContent"
                    sx="[object Object]"
                  >
                    <span
                      class="Box-sc-g0xbh4-0 mhHOW"
                      data-component="leadingVisual"
                      sx="[object Object]"
                    >
                      <svg
                        aria-hidden="true"
                        class="octicon octicon-book"
                        display="inline-block"
                        fill="currentColor"
                        focusable="false"
                        height="16"
                        overflow="visible"
                        style="vertical-align: text-bottom;"
                        viewBox="0 0 16 16"
                        width="16"
                      >
                        <path
                          d="M0 1.75A.75.75 0 0 1 .75 1h4.253c1.227 0 2.317.59 3 1.501A3.743 3.743 0 0 1 11.006 1h4.245a.75.75 0 0 1 .75.75v10.5a.75.75 0 0 1-.75.75h-4.507a2.25 2.25 0 0 0-1.591.659l-.622.621a.75.75 0 0 1-1.06 0l-.622-.621A2.25 2.25 0 0 0 5.258 13H.75a.75.75 0 0 1-.75-.75Zm7.251 10.324.004-5.073-.002-2.253A2.25 2.25 0 0 0 5.003 2.5H1.5v9h3.757a3.75 3.75 0 0 1 1.994.574ZM8.755 4.75l-.004 7.322a3.752 3.752 0 0 1 1.992-.572H14.5v-9h-3.495a2.25 2.25 0 0 0-2.25 2.25Z"
                        />
                      </svg>
                    </span>
                    <span
                      data-component="text"
                    >
                      Docs
                    </span>
                  </span>
                </button>
              </span>
              <span
                aria-label="Login"
                class="Tooltip__TooltipBase-sc-17tf59c-0 fcVRmg tooltipped-n"
                id=":r4:"
                role="tooltip"
              >
                <button
                  aria-describedby=":r5:-loading-announcement"
                  class="types__StyledButton-sc-ws60qy-0 gEWhIh"
                  data-loading="false"
                  data-testid="login-submit"
                  disabled=""
                  sx="[object Object]"
                  type="submit"
                >
                  <span
                    class="Box-sc-g0xbh4-0 srpGE"
                    data-component="buttonContent"
                    sx="[object Object]"
                  >
                    <span
                      class="Box-sc-g0xbh4-0 mhHOW"
                      data-component="leadingVisual"
                      sx="[object Object]"
                    >
                      <svg
                        aria-hidden="true"
                        class="octicon octicon-sign-in"
                        display="inline-block"
                        fill="currentColor"
                        focusable="false"
                        height="16"
                        overflow="visible"
                        style="vertical-align: text-bottom;"
                        viewBox="0 0 16 16"
                        width="16"
                      >
                        <path
                          d="M2 2.75C2 1.784 2.784 1 3.75 1h2.5a.75.75 0 0 1 0 1.5h-2.5a.25.25 0 0 0-.25.25v10.5c0 .138.112.25.25.25h2.5a.75.75 0 0 1 0 1.5h-2.5A1.75 1.75 0 0 1 2 13.25Zm6.56 4.5h5.69a.75.75 0 0 1 0 1.5H8.56l1.97 1.97a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215L6.22 8.53a.75.75 0 0 1 0-1.06l3.25-3.25a.749.749 0 0 1 1.275.326.749.749 0 0 1-.215.734Z"
                        />
                      </svg>
                    </span>
                    <span
                      data-component="text"
                    >
                      Login
                    </span>
                  </span>
                </button>
              </span>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
